### PR TITLE
Auth token background refresh

### PR DESCRIPTION
**Pull request by donfrigo**
### Description of the Change

It is important to have a valid (up to date) authentication tokens in the client side and having short-living tokens it is necessary to renew them periodically. This PR enables the automatic refresh of authTokens, before they would expire.

### Alternate Designs
Also considered updating the authToken upon getting a 401 from the server, but this approach wasn't selected as it was more complicated to implement without granting many advantages.

### Benefits
AuthTokens are refreshed automatically, so that the user can at all times query an authenticated endpoint.

### Possible Drawbacks
More stress on our server.

### Verification Process
Log in and check that the authToken is refreshed every 45 seconds (when TTL is 1 min).

### CODE REVIEW

File: src/services/auth.service.ts [hunk 1]
```diff
@@ -79,5 +80,24 @@ export const authService = {
 
     isMobile() {
         return /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
+    },
+
+    getAuthTokenExpiration(): ITokenExpiration {
+        const jwtDecode = require('jwt-decode');
```
Comment (javier-tarazaga) [inline src/services/auth.service.ts]: Don't require inline. Instead pleas use an import

File: src/epics/login/refreshAuth.epic.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import { userService, authService } from '../../services';
+import { EMPTY, of, timer } from 'rxjs';
+import { ITokenExpiration } from '../../models';
+
+export const refreshAuthStart = (action$: AnyAction, state$: any) => action$.pipe(
+    ofType(authActions.REFRESH_AUTH_START),
+    withLatestFrom(state$),
+    // currently the TTL of AuthToken is 120 sec, so refresh every 110 sec
+    map(() => authService.getAuthTokenExpiration()),
+    switchMap((JWT: ITokenExpiration) => timer(JWT.nextRefresh * 1000, JWT.refreshInterval * 1000)
```
Comment (javier-tarazaga) [inline src/epics/login/refreshAuth.epic.ts]: Don't use all capital letters to define a variable. I actually thought that was actually a constant :D

File: src/epics/login/githubLogin.epic.ts [hunk 3]
```diff
@@ -64,7 +64,7 @@ export const githubLogin = (action$: AnyAction, state$: any) => action$.pipe(
                 )),
                 switchMap((data: any) => authService.githubAuth(data)),
                 switchMap(() => userService.getUser()),
-                mergeMap((data) => [authActions.loginSuccess(data), userActions.getProjectList()]),
+                mergeMap((data) => [authActions.loginSuccess(data), authActions.refreshAuthStart(), userActions.getProjectList()]),
```
Comment (javier-tarazaga) [inline src/epics/login/githubLogin.epic.ts]: I would not perform this action here but actually when the application starts instead.

File: src/epics/login/silentLogin.epic.ts [hunk 4]
```diff
@@ -27,17 +27,17 @@ export const silentLogin = (action$: AnyAction, state$: any) => action$.pipe(
         .pipe(
             switchMap(() => userService.getUser()
                 .pipe(
-                    map(authActions.loginSuccess),
                     tap(() => console.log('Silent login success using authToken')),
+                    mergeMap((data) => [authActions.loginSuccess(data), authActions.refreshAuthStart(), userActions.getProjectList()]),
```
Comment (javier-tarazaga) [inline src/epics/login/silentLogin.epic.ts]: I would actually not perform the auth token refresh here. The idea what we talked about was to fire an event when the application started that would automatically check if it needed to refresh the token, do so if needed and set a timer (as you did)

File: src/epics/login/silentLogin.epic.ts [hunk 5]
```diff
@@ -27,17 +27,17 @@ export const silentLogin = (action$: AnyAction, state$: any) => action$.pipe(
...
                     catchError((err: any) => {
                             console.log('Unable to login using authToken: ', err.message);
                             return authService.refreshAuth()
                                 .pipe(
                                     tap((e) => console.log(e)),
                                     switchMap(() => userService.getUser()
                                         .pipe(
-                                            mergeMap((data) => [authActions.loginSuccess(data), userActions.getProjectList()]),
                                             tap(() => console.log('Silent login success using refreshToken')),
+                                            mergeMap((data) => [authActions.loginSuccess(data), authActions.refreshAuthStart(), userActions.getProjectList()]),
```
Comment (javier-tarazaga) [inline src/epics/login/silentLogin.epic.ts]: Same here, I wouldn't trigger this action manually but simply let the epic do the work

File: src/services/auth.service.ts [hunk 6]
```diff
@@ -79,5 +80,24 @@ export const authService = {
...
+        const token = getAuthToken();
+        const decoded = jwtDecode(token);
+
+        // no higher than refreshInterval in seconds
+        const refreshOffset = 15;
```
Comment (javier-tarazaga) [inline src/services/auth.service.ts]: I would actually set up that the offset as an ENV variable so we can actually control that without having to hardcode it in the source. What do you think?

File: src/services/auth.service.ts [hunk 7]
```diff
@@ -79,5 +80,24 @@ export const authService = {
...
+
+        // how many seconds before expiration do refresh
+        const TTL = decoded.exp - Math.floor(Date.now() / 1000);
+        // refresh x seconds before expiration
+        const nextRefresh = Math.max(0, TTL - refreshOffset);
+
+        // difference between expiration and issuance
+        const refreshInterval = decoded.exp - decoded.iat - refreshOffset;
```
Comment (javier-tarazaga) [inline src/services/auth.service.ts]: What is the difference between nextRefresh and refreshInterval? 🤔

File: src/services/auth.service.ts (hunk 1 above)
Comment (donfrigo) [reply src/services/auth.service.ts]: fixed

File: src/epics/login/refreshAuth.epic.ts (hunk 2 above)
Comment (donfrigo) [reply src/epics/login/refreshAuth.epic.ts]: renamed const JWT to token

File: src/services/auth.service.ts (hunk 6 above)
Comment (donfrigo) [reply src/services/auth.service.ts]: I agree, it's a better approach. Fixed.

**Pull request merged**